File: Insertion Sort/Insertion Sort.py
# This code is contributed by @MahmoudFettal

# Time Complexity: O(n^2)
# Space Complexity: O(1)

# The Insertion Sort
def insertion(L):
    n = len(L)
    for i in range(1,n):
        j = i
        value = L[i]
        while j >= 0:
            if L[j] < value:
                break
            j -= 1
        pos = j + 1
        for k in range(i,pos,-1):
            L[k] = L[k-1]
        L[pos] = value
    return L

# It is not exactly the insertion but instead of taking a position and then acting while searching for the position we move value
def insertionXbubble(L):
    n = len(L)
    for i in range(1,n):
        while i > 0:
            if L[i-1] > L[i]:
                L[i-1],L[i] = L[i],L[i-1]
            else:
                break
            i -= 1
    return L

# Test
L = [1,5,3,9,0,6,2]
print(insertion(L))

File: Insertion Sort/insertion_sort.cpp
/// This code is contributed by @iuliagroza

/// Time Complexity: O(n^2)
/// Space Complexity: O(1)

#include <iostream>
#define nMax 1005

int main()
{
    std::ios_base::sync_with_stdio(0);
    std::cin.tie(0);
    std::cout.tie(0);

    /// Input
    int n;
    std::cin >> n;

    int a[nMax];
    for(int i=1; i<=n; i++)
        std::cin >> a[i];

    /// Searching for elements that are smaller than their predecessor
    for(int i=2; i<=n; i++)
    {
        int key = a[i], j = i-1;

        /// Moving the found element on the right position
        while(j>0 && a[j]>key)
            a[j+1] = a[j--];

        a[j+1] = key;
    }

    /// Output
    for(int i=1; i<=n; i++)
        std::cout << a[i] << ' ';
}
